chore(room_preview): add `RoomListItem::preview_room`

jmartinesp opened the pull request:
This method will return a `RoomPreview` for the provided room id. 

Also added `fn RoomPreview::leave()` action to be able to decline invites or cancel knocks, since there wasn't a `Client::leave_room_by_id` counterpart as there is for  join.

The PR also deprecates `RoomListItem::invited_room`, since we have a better alternative now.

<!-- description of the changes in this PR -->

- [ ] Public API changes documented in changelogs (optional)

<!-- Sign-off, if not part of the commits -->
<!-- See CONTRIBUTING.md if you don't know what this is -->
Signed-off-by:

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_list.rs:
```diff
@@ -584,16 +584,32 @@ impl RoomListItem {
+    /// An error will be returned if the room is a state other than invited
+    /// or knocked, the 2 states that would match this use case.
+    ///
+    /// ⚠️ Holding on to this room instance after it has been joined is not
+    /// safe. Use `full_room` instead.
```
What do you mean by "not safe" here? This has quite a precise meaning in Rust, and I'm not sure if it's that, and I'm curious :)

jmartinesp replied:
This is copied from the `invited_room` fn, so we'd probably better ask @stefanceriu , but I think it means 'not safe' as in 'anything related to the timeline will fail'.

jmartinesp replied:
Resolving this as the doc comment doesn't have the 'not safe' part anymore.

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_list.rs:
```diff
@@ -584,16 +584,32 @@ impl RoomListItem {
+    ///
+    /// ⚠️ Holding on to this room instance after it has been joined is not
+    /// safe. Use `full_room` instead.
+    fn room_without_timeline(&self) -> Result<Arc<Room>, RoomListError> {
```
Really not fond of this name, I see one of three ways to go from here (there can be more!):

1. Either rename this `preview_room()` or `stripped_room()`, as it's really a very basic stripped view of a room.
2. Rename `fn room()` to `fn room_with_timeline()`, and this one can be `fn room()` later.
3. (is kind of orthogonal, and could be in addition to (1) or (2)) It might be nice to return a `StrippedRoom` or `PreviewRoom` type instead of `Room`, from this function, to make it clear which functions are available and which are not. In fact, this might even get rid of the unused internal timeline instance.

What do you think?

jmartinesp replied:
`preview_room` sounds fine to me. As for having a separate struct backing those, maybe we can create one, once we know all the values we need to expose through the FFI. I can try to get this working.

bnjbvr commented on bindings/matrix-sdk-ffi/src/lib.rs:
```diff
@@ -18,6 +18,8 @@ mod platform;
 mod room_directory_search;
 mod room_info;
+
+#[allow(deprecated)]
 mod room_list;
```
How many uses of the deprecated functions are there in this module? Can we add the guard on them one by one instead, or would that be too many? If too many, can we put the allow guard in the module itself, instead of here?

jmartinesp replied:
When I tried using the allow guard in the function it didn't take effect (I still had the same compiler error message), I had to place it in the module for it to work. I suspect it has something to do with the FFI generated code, but I'm not sure.

bnjbvr replied:
Ah yeah, expanding the macro could tell us whether it's possible to put it on a slightly more specialized location, e.g. the `impl` block itself, or if it's using it in code generated elsewhere.

jmartinesp replied:
Adding it to the impl block fails too. I've only been able to fix it by adding it in the `mod` declaration.

jmartinesp replied:
When I add it to the function, the expanded macro looks like this:

```rust
#[doc = " Builds a `Room` FFI from an invited room without initializing its"]
#[doc = " internal timeline."]
#[doc = ""]
#[doc = " An error will be returned if the room is a state different than invited."]
#[doc = ""]
#[doc = " ⚠\u{65039} Holding on to this room instance after it has been joined is not"]
#[doc = " safe. Use `full_room` instead."]
#[allow(deprecated)]
#[deprecated(note = "Please use `preview_room` instead.")]
fn invited_room(&self) -> Result<Arc<Room>, RoomListError> {
    if !match (self.membership()) {
        Membership::Invited => true,
        _ => false
    } {
        return Err(RoomListError::IncorrectRoomMembership {
            expected: (<[_]>::into_vec(
                #[rustc_box]
                ::alloc::boxed::Box::new([Membership::Invited])
            )),
            actual: self.membership(),
        });
    }
    Ok(Arc::new(Room::new(self.inner.inner_room().clone())))
}
```

And if I add it to the impl block it looks like:

```rust
#[uniffi::export(async_runtime = "tokio", )]
#[allow(deprecated)]
impl RoomListItem {
    ...
```

But clippy still complains in both cases.

bnjbvr replied:
Yeah, it's the call sites that need the `allow(deprecated)` guard, not the function itself; but it might be that it's in code generated by uniffi, that's outside our control, so we might need to resort to the mod-level guard indeed.

bnjbvr replied:
Forgot to comment here: it's indeed an unmangled free-function generated in the same module, that's calling the `invited_room` method, so welp :shrug: Module-level guard it is. I wondered if we could get rid of the entire `invited_room` method altogether, but I don't want to add more API changes to the other side of this PR.

Hywan commented on bindings/matrix-sdk-ffi/src/room_list.rs:
```diff
@@ -584,16 +584,32 @@ impl RoomListItem {
+    /// safe. Use `full_room` instead.
+    fn room_without_timeline(&self) -> Result<Arc<Room>, RoomListError> {
+        let membership = self.membership();
+        if !matches!(membership, Membership::Invited)
+            && !matches!(self.membership(), Membership::Knocked)
```
It can be a single `matches!`.

```suggestion
        if !matches(membership, Membership::Invited | Membership::Knocked)
```

jmartinesp commented:
I ended up creating a `RoomPreviewActions` struct in the FFI layer so when you request a `RoomListItem::preview_room` you get a `RoomPreview` record that contains `RoomPreviewInfo` (also a record) and `RoomPreviewActions`, which are either `join`, `leave` or `knock` and check the membership state of the room preview before running. So from a client, you'd do something like:

```kotlin
val preview = roomListItem.previewRoom()
if (preview.info.membership == Membership.Invite) {
    preview.actions.join()
}
```

bnjbvr commented on crates/matrix-sdk/src/room_preview.rs:
```diff
@@ -219,6 +250,105 @@ impl RoomPreview {
     }
 }
+
+pub enum RoomPreviewActions {
```
In this entire file you need to comment each struct, enum, variant, field, function. I'll enable a new lint for this :smiling_imp:

jmartinesp replied:
This was fixed in the following commit if I'm not mistaken.

bnjbvr commented on crates/matrix-sdk/src/room_preview.rs:
```diff
@@ -219,6 +250,105 @@ impl RoomPreview {
+    WrongRoomPreviewState(WrongRoomPreviewState),
+
+    #[error(transparent)]
+    Generic(#[from] Box<dyn std::error::Error + Send + Sync>)
```
I really don't like this; having an exceptional generic error type like that should be limited at the strict minimum, i.e. `matrix_sdk::Error` only. Let's avoid it entirely by not making a custom error, but instead use an `matrix_sdk::Error` instead of this error type, there's already `WrongRoomState` which describes the same error.

jmartinesp replied:
Removed in https://github.com/matrix-org/matrix-rust-sdk/commit/0fcfd84311e2047825763aeae41cc210d3007041.

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_list.rs:
```diff
@@ -580,31 +584,47 @@ impl RoomListItem {
 
-    /// Builds a `Room` FFI from a room without initializing its internal
-    /// timeline.
+    /// Builds a `RoomPreview` from a room list item. This is intended for rooms
+    /// with [`Membership::Invite`] or [`Membership::Knocked`],
```
nit: this sentence ends with a comma, and it doesn't really translate to proper English IMO: maybe `This is intended for invited or knocked rooms.`?

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_preview.rs:
```diff
@@ -1,9 +1,87 @@
+
+struct
+
+impl RoomPreview {
```
Seems like this is not an atomic commit… (It's fine but next time to make the review easier, feel free to squash them altogether, since I know that another commit will now change these three lines.)

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_preview.rs:
```diff
@@ -1,9 +1,87 @@
+
 #[derive(uniffi::Record)]
 pub struct RoomPreview {
+    pub info: RoomPreviewInfo,
```
Why do we need the intermediary `RoomPreviewInfo` data structure?

jmartinesp replied:
So we didn't need to move everything inside `RoomPreview` into a `Record` if we made it an `Object` (as is the case now since I'm removing the separate `actions` part).

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_preview.rs:
```diff
@@ -1,9 +1,87 @@
+#[matrix_sdk_ffi_macros::export]
+impl JoinRoomPreviewAction {
+    async fn run(&self) {
+        self.inner.run().unwrap()
```
Let's not `unwrap()` at the FFI layer, this will cause unnecessary panics that we need to handle at the last minute.

jmartinesp replied:
Removed in https://github.com/matrix-org/matrix-rust-sdk/pull/4152/commits/0fcfd84311e2047825763aeae41cc210d3007041.

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_preview.rs:
```diff
@@ -1,9 +1,87 @@
+            num_joined_members: info.num_joined_members,
+            room_type: info.room_type.map(|room_type| room_type.to_string()),
+            is_history_world_readable: info.is_world_readable,
+            membership: info.state.map(|state| state.into()).unwrap_or_else(|| Membership::Left),
```
nit: `unwrap_or` is sufficient; `unwrap_or_else` is to be used when some computation needs to happen in the closure.

Why can we assume `Left` in the first place, though?

jmartinesp replied:
That's what `RoomPreview` does at the SDK level [here](https://github.com/matrix-org/matrix-rust-sdk/blob/b6e9d4f6bdb35b4f5c5320c63133b0a33d278c21/crates/matrix-sdk/src/room_preview.rs#L114) for join rules, so I assumed we could do the same. I'll change it to `Option<Membership>` instead.

jmartinesp replied:
Handled in https://github.com/matrix-org/matrix-rust-sdk/commit/7bfb1bd1c4d3bd65f15b132330f3dbc9189063f0.

bnjbvr commented on crates/matrix-sdk/src/room_preview.rs:
```diff
@@ -69,18 +77,27 @@ pub struct RoomPreview {
     /// we can do better than that.
-    fn from_room_info(
+    async fn from_room_info(
+        client: &Client,
```
nit: since both callers have a handle to the `room` from the Client, can you pass `is_direct: bool` here instead?

jmartinesp replied:
Sure, but I'd still have to call `room.is_direct().await.unwrap_or_default()`, is it fine to assume it's not direct at this point?

jmartinesp replied:
I'd still get an `Option<bool>` in some cases. Maybe I'll just pass the nullable `is_direct` field to the clients and let them decide what to do with it.

jmartinesp replied:
I'll still need to pass a `Client` around since I need it to be able to call `RoomPreview::leave()`, as there's no other way to leave an invited/knocked room if we change those to return a `RoomPreview` instead. Should I still calculate the `is_direct` value outside, if the client can do it inside with the latest info?

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_list.rs:
```diff
@@ -624,7 +622,7 @@ impl RoomListItem {
 
         let room_preview = client.get_room_preview(&room_or_alias_id, server_names).await?;
-        Ok(Arc::new(RoomPreview::try_from_sdk(room_preview, client)))
+        RoomPreview::try_from_sdk(room_preview, client)
```
This fixes stuff from the previous commit, right? :pensive:

bnjbvr commented on crates/matrix-sdk/src/room_preview.rs:
```diff
@@ -254,101 +250,181 @@ impl RoomPreview {
+    pub fn new(client: Client, room_preview: RoomPreview) -> Self {
+        let room_id = room_preview.room_id;
+        let room_state = room_preview.state;
+        let is_direct = room_preview.is_direct.unwrap_or_default();
```
That's a lie, let's keep an `Option<bool>` for `is_direct` ;)

jmartinesp replied:
This code was removed.

bnjbvr commented on crates/matrix-sdk/src/room_preview.rs:
```diff
@@ -254,101 +250,181 @@ impl RoomPreview {
-    Knocked { leave: LeaveRoomPreviewAction },
+/// A set of actions that can be performed in a [`RoomPreview`].
+#[derive(Debug)]
+pub struct RoomPreviewActions {
```
Overall, I don't think this data structure and the other actions are required at the SDK level; maybe they're required at the FFI layer because we need some kind of "lazy provider"-like object for the Record requirements, but I don't think they make sense for the `RoomPreview` of the SDK, considering there are many other methods that already exist providing the same functionality, so removing it would avoid blowing up the number of abstractions, code duplication, etc.

jmartinesp replied:
I added it to the SDK so they could be tested and didn't have to duplicate code, i.e. the `Room::join` contents mentioned below, which could be just put into one of these actions and then reused inside `Room::join` instead:

```rust
impl Room {
    async fn join(&self) -> Result<_, _> {
        self.actions.join()
   }
```

But if that's not something we want, we can just remove the actions, leaving a needed `RoomPreview::leave` method that's needed (`Client` has code for joining and knocking, but it doesn't have one for leaving a room given its id, so a `RoomPreview` with invite/knock state can never be left like that).

bnjbvr reviewed (changes requested):
I think we don't need all the preview actions data structure at the SDK level; these need to exist at the FFI layer because of… reasons (that I'm detailed below; let me know if that's wrong). In the current shape I am strongly not in favor of merging this, let's identify what's the minimal set of changes that we could have to unblock us, and maybe consider doing some refactoring later.

jmartinesp commented on bindings/matrix-sdk-ffi/src/room_preview.rs:
```diff
@@ -21,33 +99,22 @@ pub struct RoomPreview {
+            SpaceRoomJoinRule::Knock => JoinRule::Knock,
+            SpaceRoomJoinRule::KnockRestricted => JoinRule::KnockRestricted { rules: Vec::new() },
+            SpaceRoomJoinRule::Restricted => JoinRule::Restricted { rules: Vec::new() },
+            // For the `_Custom` case, assume it's private
+            _ => JoinRule::Private,
```
I'll add a `Custom(String)` variant, although I guess in the clients we'll just need to assume it's private anyway.

jmartinesp replied:
We don't have a way to map the `PrivOwnedStr` in `_Custom` to a `String` AFAICT, or to create a `PrivOwnedStr` from a String either. I think this is why the mappers just assumed the `Private` usage.

jmartinesp commented:
- I removed the `Room*Actions` in https://github.com/matrix-org/matrix-rust-sdk/pull/4152/commits/0fcfd84311e2047825763aeae41cc210d3007041.
- I fixed the `RoomPreview` contents and added a `fn RoomPreview::leave()` that was needed in https://github.com/matrix-org/matrix-rust-sdk/pull/4152/commits/7bfb1bd1c4d3bd65f15b132330f3dbc9189063f0.
- The doc comment using direct references to types was fixed in https://github.com/matrix-org/matrix-rust-sdk/pull/4152/commits/f8110d24e28b12e6f7a31fe14e1cad2551da6165.
- I tried to explain why the `JoinRule::_Custom` case can't be handled and we default to `Private` in https://github.com/matrix-org/matrix-rust-sdk/pull/4152/commits/69ee5c2e4defb4ad3b6400f2fa7246d6ae4ad14d.

bnjbvr commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -1208,6 +1208,13 @@ impl Client {
         Ok(Room::new(self.clone(), base_room))
     }
 
+    /// Leave a room given its `RoomId`.
```
We don't need this method, it's the same as `Room::leave()`, and we can only leave a room we do know about (i.e. we can get it with `Client::get_room` first).

jmartinesp replied:
I think we're going in circles... We don't have anything in the `ffi::Client` to get a `Room`, and even if we had it, it would be a 'not safe to use' Room as we had in `RoomListItem::invited_room` (see https://github.com/matrix-org/matrix-rust-sdk/pull/4152#discussion_r1808445972). That's the main reason I split Room and RoomPreview here, and I think it's what we agreed to do.

So now I'm not sure what to do anymore:
- If we're ok with returning a Room with no timeline as we did before then I'll just remove everything from `RoomPreview` and just return that Room in `RoomListItem::preview_room` instead.
- If we're not ok with that, then we need a way to be able to leave that room (aka decline the invite, cancel the knock).

bnjbvr commented on crates/matrix-sdk/src/room_preview.rs:
```diff
@@ -69,19 +69,46 @@ pub struct RoomPreview {
 }
 
 impl RoomPreview {
+    /// Leave a room using its [`RoomPreview`].
+    pub async fn leave(&self) -> Result<(), Error> {
+        match self.state {
+            Some(RoomState::Left) | None => {
+                Err(Error::WrongRoomPreviewState(WrongRoomPreviewState {
+                    expected: "Invited, Knocked or Joined",
+                    got: self.state,
+                }))
+            }
+            Some(_) => self.client.leave_room(&self.room_id).await.map_err(Into::into),
+        }
+    }
```
Sorry, I still think that `RoomPreview` ought to be a purely descriptive data record (i.e. no action methods); can the caller use the room id that's always around, and call `leave_room` by themselves? Then, can we get rid of the `Client` stored and address the other comment I had on `from_room_info`?

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_preview.rs:
```diff
@@ -20,34 +65,28 @@ pub struct RoomPreview {
+                // Warning! If new join rules are introduced we may be mapping wrong values
+                // here, but there's no way to match
+                // `SpaceRoomJoinRule::_Custom(_)` and have an exhaustive match.
+                JoinRule::Private
```
I'd do it differently:

- for custom, add a `JoinRule::Custom(String)` variant that's built with `join_rule.to_string()` (`SpaceRoomJoinRule` implements `StringEnum` which derives a `Display` impl)?
- for `_`, log a warning and return `None`. Yes, let's turn this into `TryInto` :innocent:

bnjbvr commented on bindings/matrix-sdk-ffi/src/client.rs:
```diff
@@ -1027,7 +1027,7 @@ impl Client {
         room_id: String,
         via_servers: Vec<String>,
-    ) -> Result<RoomPreview, ClientError> {
+    ) -> Result<Arc<RoomPreview>, ClientError> {
```
Do we still need all the `Arc`s here? (I'm not very sure about it myself, but I suspect we don't since `RoomPreview` is now a uniffi object)

bnjbvr commented on bindings/matrix-sdk-ffi/src/room_list.rs:
```diff
@@ -574,31 +576,63 @@ impl RoomListItem {
+            (room_or_alias_id, Vec::new())
+        } else {
+            let room_or_alias_id: OwnedRoomOrAliasId = self.inner.id().to_owned().into();
+            let server_names: Vec<OwnedServerName> = via
+                .into_iter()
+                .map(|server| ServerName::parse(server).map_err(ClientError::from))
+                .collect::<Result<_, ClientError>>()?;
```
It's a bit weird that if there's a canonical alias, the `via` routes will not be validated :thinking: Should we validate them ahead, even if we're discarding them before?

bnjbvr commented on crates/matrix-sdk/tests/integration/room_preview.rs:
```diff
@@ -0,0 +1,139 @@
+}
+
+#[async_test]
+async fn test_room_preview_leave_knocked() {
```
With the suggestions I've made above, I think that most of these tests could go away. Unless we didn't have tests to leave a knocked room, maybe?

bnjbvr reviewed (changes requested):
Sorry, it's still too many unneeded changes for me. Let me try to hack on this, see if I can get it into the shape I'm expecting.

jmartinesp commented on bindings/matrix-sdk-ffi/src/room_preview.rs:
```diff
@@ -20,34 +89,32 @@ pub struct RoomPreview {
+            SpaceRoomJoinRule::Public => JoinRule::Public,
+            SpaceRoomJoinRule::_Custom(_) => JoinRule::Custom { repr: join_rule.to_string() },
+            _ => {
+                // Since we have no way to get the _Custom contents, assume it's private.
+                // Warning! If new join rules are introduced we may be mapping wrong values
+                // here, but there's no way to match
+                // `SpaceRoomJoinRule::_Custom(_)` and have an exhaustive match.
```
Maybe remove/fix this comment?

bnjbvr commented:
Jorge confirmed this was good enough for the EX apps, so merging. Thanks @jmartinesp for all the work done here, and sorry about the few back-and-forths.
